lib/linked_list.rb

```diff
@@ -1,45 +1,89 @@
-require_relative 'node'
+require_relative "node"
 
 class LinkedList
   attr_reader :head
+  attr_accessor :length, :tail
 
   def initialize
     @head = nil
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(1)
+  # Space complexity - O(1)
   def add_first(data)
-
+    @head = Node.new(data, head)
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(1)
+  # Space complexity - O(1)
   def get_first
-
+    if head.nil?
+      return nil
+    end
+    return head.data
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(1)
+  # Space complexity - O(1)
   def length
-    return 0
+    length = 0
+    current = head
+    if head.nil?
+      return 0
+    end
+    until current.nil?
+      length += 1
+      current = current.next
+    end
+    return length
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(1)
+  # Space complexity - O(1)
   def add_last(data)
+    if head.nil?
+      add_first(data)
+      return head.data
+    end
+
+    current = head
 
+    while !current.next.nil?
+      current = current.next
+    end
+
+    new_last = Node.new(data)
+    current.next = new_last
+    return new_last.data
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(1)
+  # Space complexity - O(1)
   def get_last
-
+    current = head
```

Comment:
What if `head` is `nil`?

```suggestion
    return nil if head.nil?
    current = head
```